File: backend/app/domain/forecast_runs.py
"""Auditable stored forecast-run domain types."""

from dataclasses import dataclass
from datetime import date, datetime
from typing import Literal

RunStatus = Literal["completed", "partial", "failed"]
SeriesRunStatus = Literal["succeeded", "failed"]
EvaluationStatus = Literal["succeeded", "failed"]


@dataclass(frozen=True, slots=True)
class ForecastRun:
    run_id: str
    data_version: str
    status: RunStatus
    created_at: datetime
    training_cutoff: date
    horizon_months: int
    interval_level: int
    series_count: int
    successful_series_count: int
    failed_series_count: int
    evaluation_count: int
    forecast_count: int


@dataclass(frozen=True, slots=True)
class ForecastSeriesResult:
    run_id: str
    store_id: str
    product_id: str
    sku: str
    product_name: str
    category: str
    status: SeriesRunStatus
    selected_model: str | None
    wape: float | None
    mase: float | None
    rmse: float | None
    bias: float | None
    interval_coverage: float | None
    validation_points: int | None
    failure_detail: str | None


@dataclass(frozen=True, slots=True)
class StoredModelEvaluation:
    run_id: str
    store_id: str
    product_id: str
    sku: str
    model_name: str
    selected: bool
    status: EvaluationStatus
    validation_origins: int
    wape: float | None
    mase: float | None
    rmse: float | None
    bias: float | None
    interval_coverage: float | None
    validation_points: int | None
    failure_detail: str | None


@dataclass(frozen=True, slots=True)
class StoredForecast:
    run_id: str
    store_id: str
    product_id: str
    sku: str
    period_start: date
    forecast_value: float
    lower_bound: float
    upper_bound: float
    horizon: int
    selected_model: str
    training_cutoff: date

File: frontend/src/api/forecast-runs.ts
import { getJson } from './http-client';

export interface ForecastRun {
  readonly run_id: string;
  readonly data_version: string;
  readonly status: 'completed' | 'partial' | 'failed';
  readonly created_at: string;
  readonly training_cutoff: string;
  readonly horizon_months: number;
  readonly interval_level: number;
  readonly series_count: number;
  readonly successful_series_count: number;
  readonly failed_series_count: number;
  readonly evaluation_count: number;
  readonly forecast_count: number;
}

export interface ForecastSeriesResult {
  readonly run_id: string;
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly product_name: string;
  readonly category: string;
  readonly status: 'succeeded' | 'failed';
  readonly selected_model: string | null;
  readonly wape: number | null;
  readonly mase: number | null;
  readonly rmse: number | null;
  readonly bias: number | null;
  readonly interval_coverage: number | null;
  readonly validation_points: number | null;
  readonly failure_detail: string | null;
}

export interface StoredModelEvaluation {
  readonly run_id: string;
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly model_name: string;
  readonly selected: boolean;
  readonly status: 'succeeded' | 'failed';
  readonly validation_origins: number;
  readonly wape: number | null;
  readonly mase: number | null;
  readonly rmse: number | null;
  readonly bias: number | null;
  readonly interval_coverage: number | null;
  readonly validation_points: number | null;
  readonly failure_detail: string | null;
}

export interface StoredForecast {
  readonly run_id: string;
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly period_start: string;
  readonly forecast_value: number;
  readonly lower_bound: number;
  readonly upper_bound: number;
  readonly horizon: number;
  readonly selected_model: string;
  readonly training_cutoff: string;
}

export function listForecastRuns(signal?: AbortSignal): Promise<ForecastRun[]> {
  return getJson<ForecastRun[]>('/api/v1/forecast-runs', signal);
}

export function listForecastSeries(
  runId: string,
  storeId: string,
  query: string,
  signal?: AbortSignal,
): Promise<ForecastSeriesResult[]> {
  const parameters = new URLSearchParams({ query, limit: '100' });
  if (storeId) parameters.set('store_id', storeId);
  return getJson<ForecastSeriesResult[]>(
    `/api/v1/forecast-runs/${encodeURIComponent(runId)}/series?${parameters.toString()}`,
    signal,
  );
}

export function getStoredEvaluations(
  runId: string,
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<StoredModelEvaluation[]> {
  const parameters = new URLSearchParams({ store_id: storeId, product_id: productId });
  return getJson<StoredModelEvaluation[]>(
    `/api/v1/forecast-runs/${encodeURIComponent(runId)}/evaluations?${parameters.toString()}`,
    signal,
  );
}

export function getStoredForecasts(
  runId: string,
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<StoredForecast[]> {
  const parameters = new URLSearchParams({
    run_id: runId,
    store_id: storeId,
    product_id: productId,
  });
  return getJson<StoredForecast[]>(`/api/v1/forecasts?${parameters.toString()}`, signal);
}
